### ISSUE TITLE
concatenate_videoclips without re-encoding

### ISSUE THREAD

**Issue by assafmendelson** [Zulko/moviepy#737]
Consider a basic clip concatenation:

    concatClips = concatenate_videoclips([VideoFileClip(x) for x in clipsFileNames])
    concatClips.write_videofile(target)

When doing this, the file is reencoded. Is there a way to combine the clips without re encoding everything?

**Comment by mgaitan** [comment]
this is very interesting topic. AFAIK, this isn't possible due the way moviepy uses ffmpeg. 

Basically, clips works reducing everything to "rawvideo" frames, and `write_videofile` just iter every frame on the clip and pipe them to the "target" ffmpeg command to make the ouput video with its own properties. 

ffmpeg, however, is capable to concatenate without reencoding. For instance, using the "demuxer mode". https://trac.ffmpeg.org/wiki/Concatenate#demuxer  

The problem is that ffmpeg's "concat" modes are targeted to uses files (of the same type) directly, not frame by frame as moviepy works. So, for your very specific example, would be possible (and more or less easy) to make a new function that receives the list of `VideoFileClip` instances (as they keep a reference to the original filename) and concatenate them using ffmpeg with no reencoding, but that wouldn't work if a clip is a subclip or a non-from-file clip.

**Comment by keikoro** [comment]
Can this be closed?

**Comment by CMCDragonkai** [comment]
Is there a way to not re-encode the videos when just extracting subclips?

**Comment by JanPokorny** [comment]
@CMCDragonkai Yes: https://stackoverflow.com/a/37323543/6099426

**Comment by tburrows13** [comment]
> Can this be closed?

I'd like to suggest a `ffmpeg_concatenate()` function in `moviepy.video.io.ffmpeg_tools` that is a simple wrapper around ffmpeg for fast concatenation.

**Comment by JanPokorny** [comment]
@tburrows13 Why not use a dedicated ffmpeg library for this type of task? https://kkroening.github.io/ffmpeg-python/

**Comment by CMCDragonkai** [comment]
@JanPokorny that takes time inputs. Can that take frame count inputs?

**Comment by JanPokorny** [comment]
@CMCDragonkai Since ffmpeg works with constant frame rate (https://superuser.com/a/1512583/662133), you should be able to simply divide frame count input by FPS to get time input.

**Comment by tburrows13** [comment]
> @tburrows13 Why not use a dedicated ffmpeg library for this type of task? https://kkroening.github.io/ffmpeg-python/

Yes, that can absolutely be used for more complex ffmpeg things. I was just suggesting a simple single function for a common use case.

**Comment by JanPokorny** [comment]
@tburrows13 Seems like feature creep tbh, since it's completely aside from the main library and quite easily doable with specialized libraries... But again, that could be said about the whole `moviepy.video.io.ffmpeg_tools`.

**Comment by OsaAjani** [comment]
Thank you for your contributions and for reporting issues in this repository. With the release of v2, which introduces significant changes to the codebase and API, we’ve reviewed the backlog of open PRs and issues. Due to the length of the backlog and the likelihood that many of these are either fixed or no longer applicable, we’ve made the decision to close all previous PRs and issues.

If you believe that any of these are still relevant to the current version or if you'd like to reopen a related discussion, please feel free to create a new issue or pull request, referencing the old one.

Thank you for your understanding and continued support!
